### PR TITLE
bpo-35993: Fix an incorrect use of released memory

### PR DESCRIPTION
**Pull request by matrixise**
https://bugs.python.org/issue35993

### CODE REVIEW

File: Python/pystate.c
Comment (vstinner) [inline Python/pystate.c]: I don't understand how the current code can work since PyMem_RawFree() fills freed memory with a byte pattern 0xDB and so interp = interp->next should just crash.

I don't see how this change fix anything: interp still points to freed bytes.

You should read interp->next before PyMem_RawFree(), and so change how the loop is written.

cc @ericsnowcurrently

Comment (eamanu) [reply Python/pystate.c]: When  ```PyMem_RawFree(tmp_interp)``` the ```tmp_interp```  point to ```interp```, so you are freed ```interp``` too, right?

What about this?
```C
tmp_interp = interp->next;
for (; tmp_interp !=NULL; interp = tmp_interp){
...
tmp_interp = interp->next;
PyMem_RawFree(interp);
}

```

Comment (matrixise) [reply Python/pystate.c]: sure? the last time I have coded in C, that was in 2007 :/ ok

Comment (matrixise) [reply Python/pystate.c]: or just
```c
PyInterpreterState *tmp_interp = NULL;
for (; interp != NULL;) {
    ...
    tmp_interp = interp;
    interp = interp->next;
    PyMem_RawFree(tmp_interp);
}
```
@vstinner and @eamanu

Comment (matrixise) [reply Python/pystate.c]: @eamanu are you on Zulip or IRC?

Comment (eamanu) [reply Python/pystate.c]: @matrixise I am connecting to zulip

Comment (JulienPalard) [reply Python/pystate.c]: With a `while (interp)` should work the same, and the jungling of tmp/interp looks good to me.

Comment (matrixise) [reply Python/pystate.c]: @JulienPalard I have just updated my PR, thanks for the review.
Now, I am going to wait for the review of @vstinner, he is more expert than me with the C language.

**Review by JulienPalard — DISMISSED** [review]
LGTM

**Review by eamanu — APPROVED** [review]
lgtm

File: Python/pystate.c
Comment (vstinner) [inline Python/pystate.c]: Remove debug printf.

File: Python/pystate.c
Comment (vstinner) [inline Python/pystate.c]: Please don't add commented code.

File: Python/pystate.c
Comment (vstinner) [inline Python/pystate.c]: You can declare "PyInterpreterState *tmp_interp = NULL;" here. I suggest to rename it "prev_interp".

**Review by vstinner — CHANGES_REQUESTED** [review]

**Review by vstinner — COMMENTED** [review]
I don't understand how the current code can code: it must crash if I understand correctly.

Is the function called? You may modify the code to add abort(), run the full Python test suite and check if a test does crash.

If there is no crash, would it be possible to add a test?

cc @ericsnowcurrently

**Comment by vstinner** [comment]
```
for (; tmp_interp !=NULL; interp = tmp_interp){
...
tmp_interp = interp->next;
PyMem_RawFree(interp);
}
```

I prefer this version, especially if you rename tmp_interp to next_interp.

**Comment by JulienPalard** [comment]
Current version does not compile, runs indefinitely, making CI timeout.

File: Python/pystate.c
Comment (JulienPalard) [reply Python/pystate.c]: (Those have been pushed recently to diagnose the "take infinite time to build when this code is added" / "make CI timeout" issue)

Comment (matrixise) [reply Python/pystate.c]: @vstinner yep, just for the debug and because I wanted to show you

File: Python/pystate.c
Comment (matrixise) [reply Python/pystate.c]: yep, sure this PR is in marked WIP and I wanted to have your feedback about the code.

**Comment by matrixise** [comment]
@vstinner it's my fault if my PR is broken after the approval of @JulienPalard. But I am working on it and try to fix it.

**Comment by matrixise** [comment]
@vstinner @ericsnowcurrently the current code is used when you compile Python, but there is no crash (un/fortunately)

**Review by JulienPalard — APPROVED** [review]
LGTM

**Comment by matrixise** [comment]
@vstinner the code works and the tests pass. no lock on travis.

**Comment by vstinner** [comment]
> @vstinner @ericsnowcurrently the current code is used when you compile Python

Would you mind to elaborate? I would like to make sure that the function is properly tested.

> but there is no crash (un/fortunately)

Strange. So maybe the current code is just fine? Or we miss a test.

**Comment by matrixise** [comment]
@vstinner example when you compile python
```
gcc -pthread -shared build/temp.linux-x86_64-3.8-pydebug/home/stephane/src/github.com/python/cpython/Modules/_contextvarsmodule.o -L/home/stephane/src/github.com/python/cpython-add_smtpshandler/lib -L/usr/local/lib -o build/lib.linux-x86_64-3.8-pydebug/_contextvars.cpython-38dm-x86_64-linux-gnu.so
1 _PyInterpreterState_DeleteExceptMain
2 _PyInterpreterState_DeleteExceptMain
#### _PyInterpreterState_DeleteExceptMain
building 'cmath' extension
gcc -pthread -fPIC -Wno-unused-result -Wsign-compare -g -Og -Wall -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Werror=implicit-function-declaration -I./Include/internal -I./Include -I/home/stephane/src/github.com/python/cpython-add_smtpshandler/include -I. -I/usr/local/include -I/home/stephane/src/github.com/python/cpython/Include -I/home/stephane/src/github.com/python/cpython -c /home/stephane/src/github.com/python/cpython/Modules/cmathmodule.c -o build/temp.linux-x86_64-3.8-pydebug/home/stephane/src/github.com/python/cpython/Modules/cmathmodule.o
1 _PyInterpreterState_DeleteExceptMain
2 _PyInterpreterState_DeleteExceptMain
#### _PyInterpreterState_DeleteExceptMain

```

**Comment by vstinner** [comment]
Ok, but I guess that the code path which trigger the bug is not taken, since according to the bug tracker, it must crash. I guess that no test call this function with more than 1 interpreter. That's why I added @ericsnowcurrently in copy of this change :-)

**Comment by eamanu** [comment]
I am not sure when `_PyInterpreterState_DeleteExceptMain()` is called. This way we could test it. @vstinner @matrixise

File: Python/pystate.c
Comment (methane) [inline Python/pystate.c]:
```suggestion
        PyInterpreterState *prev_interp = interp;
```

Please define variable where it is initialized, as possible.

Comment (matrixise) [reply Python/pystate.c]: @methane and about the compiler, this variable will be created on the stack, ok, but for each time or just one time?

**Comment by matrixise** [comment]
@eamanu as you can see my paste in the previous comment, the function is called when we try to compile Python, certainly in the Makefile.

File: Python/pystate.c
Comment (methane) [reply Python/pystate.c]: One (or zero) time.

**Review by methane — APPROVED** [review]

**Comment by methane** [comment]
> I don't understand how the current code can code: it must crash if I understand correctly.

@vstinner this for loop iterate over (sub)interpreters.
When there is only main interpreter, `PyMem_RawFree(interp);` is never called.

https://github.com/python/cpython/blob/3e028b2d40370dc986b6f3146a7ae927bc119f97/Python/pystate.c#L285-L289

**Comment by matrixise** [comment]
@ericsnowcurrently Any advice for a test for this part?
Thank you,

File: Misc/NEWS.d/next/Core and Builtins/2019-02-14-09-17-54.bpo-35993.Bvm3fP.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2019-02-14-09-17-54.bpo-35993.Bvm3fP.rst]: Only from this NEWS entry, it's hard to know which code is impacted. I found PyOS_AfterFork_Child() which calls _PyInterpreterState_DeleteExceptMain(). I understand that the bug triggers if an application using multiple interpreters forks... it sounds like a very unusual use case. I propose:

"Fix a crash on fork when using subinterpreters. Contributed by Stéphane Wirtel."

**Comment by matrixise** [comment]
@vstinner I have just updated the blurb entry, sorry I didn't see your notification

**Comment by vstinner** [comment]
Ah, it's a new function in Python 3.8. I didn't know. No need to backport in that case.

**Comment by matrixise** [comment]
@vstinner thank you for the merge ;-) a good news for today ;-)

**Pull request merged**
